File: jlc_clojure/src/fundingcircle/jukebox/client/step_registry.clj
(ns fundingcircle.jukebox.client.step-registry
  "Step registry."
  (:require [clojure.tools.logging :as log])
  (:import (java.util UUID)))

(defn create
  "Creates a step registry."
  []
  {:definitions [] :callbacks {}})

(defn definitions
  "Gets the step definitions that have been loaded."
  [step-registry]
  (:definitions step-registry))

(defn callbacks
  "Gets the step callbacks that have been found."
  [step-registry]
  (:callbacks step-registry))

(defn entry-points
  "The list of entry point functions."
  [step-registry]
  (vals (callbacks step-registry)))

(defn add
  "Add a step or hook to the step registry."
  [step-registry {:keys [triggers opts callback]}]
  (let [id (str (UUID/randomUUID))
        tags (:scene/tags opts)
        tags (if (string? tags) [tags] tags)
        opts (dissoc (assoc opts :scene/tags tags) :tags)]
    (-> step-registry
        (update :callbacks assoc id callback)
        (update :definitions conj {:id id :triggers triggers :opts opts}))))

(defn run
  "Run a step or hook."
  [step-registry {:keys [id board args] :as message}]
  (let [callback (get-in step-registry [:callbacks id])]
    (when-not callback (throw (ex-info "Undefined callback" {:message message})))
    (log/debugf "Running step %s: %s" id {:board board :args args})
    (apply callback board args)))


(defn find-trigger
  "Finds the step definition for the trigger."
  [step-registry trigger]
  (first (filter #(some (fn [t] (= trigger t)) (:triggers %)) (:definitions step-registry))))

File: jlc_ruby/lib/jukebox/client/step_registry.rb
# frozen_string_literal: true

require 'active_support/core_ext'
require 'logger'
require 'singleton'
require 'securerandom'
require 'set'

module Jukebox
  class Client
    # Step registry.
    class StepRegistry
      include Singleton
      attr_reader :definitions, :callbacks

      def initialize
        @definitions = []
        @callbacks = {}
        @resources = Set[]
      end

      # Add a step or hook to the step registry.
      def add(*triggers, **opts, &block)
        raise UndefinedError unless block

        triggers.map! { |t| t.is_a?(Symbol) ? t : t.inspect[1..-2] }
        id = SecureRandom.uuid
        @definitions << {
          id: id,
          triggers: triggers,
          opts: StepRegistry.cleanup_opts(opts)
        }
        @callbacks[id] = block
        @resources += opts[:resources] if opts[:resources]
      end

      # Run a step or a hook
      def run(id:, board:, args:, **message)
        callback = @callbacks[id]
        raise UndefinedError, "Undefined callback: #{message}" unless callback

        callback.call(board, *args)
      end

      # Finds a step's definition
      def find_definition(trigger)
        @definitions.find { |d| d[:triggers].include? trigger }
      end

      # Normalizes a step's options
      def self.cleanup_opts(opts)
        opts['scene/tags'.to_sym] = [opts[:tags]] if opts[:tags].is_a?(String)
        opts['scene/resources'.to_sym] = opts[:resources] if opts[:resources]
        opts.except!(:tags, :resources)
      end
    end
  end
end
